match leading slash imports with path mappings - fixes #13730

EECOLOR opened the pull request:
Checklist:
* [x] There is an [associated issue](https://github.com/Microsoft/TypeScript/issues/13730)
* [ ] The issue is labeled 'Bug' or 'help wanted' or is in the Community milestone
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `jake runtests` locally
* [x] You've signed the CLA
* [x] There are new or updated unit tests validating the change

About the code changes: I can imagine you would like to have the code structured differently or have different traces. I just looked at what currently been implemented and took the most straightforward approach making the changes easy to diff. Feel free to make suggestions.

About the tests: I am missing information to provide the correct amount of tests, or actually I don't know what your policy is. I think the tests I added should cover all of the code I added and prevent future breakage. 

About performance: I tried to minimize the performance impact. I am not sure how often this code is called in situations where high performance is required. That said, I noticed a lot of places where performance could be improved, so I am not sure how much you guys want to focus on that.

About caching: I noticed a few places where `isExternalModuleNameRelative` is used in relation to caching. We might want to add the 'is potentially an alias import' there as well. This however requires some extra thought. An example of such a place:

```js
if (perFolderCache) {
  perFolderCache.set(moduleName, result);
  if (!isExternalModuleNameRelative(moduleName)) {
    // put result in per-module name cache
    cache!.getOrCreateCacheForModuleName(moduleName).set(containingDirectory, result);
  }
}
```

Fixes #13730

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=27980) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=27980) before we can accept your contribution.<br/><br/>:x: EECOLOR [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=27980)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=27980) it.</sub>

ajafff commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -1433,3 +1444,9 @@ namespace ts {
 }
+/* @internal */
+namespace ts {
+    export function startsWithSlash(name: string) {
```
Since this function is not used anywhere else, you could avoid exporting it

EECOLOR replied:
Great tip. I tried that first, but I had some trouble which seemed to be caused by the `/* internal */` comment, it could not resolve `CharacterCodes`. I just tried to move it again, and now it works. My knowledge of TypeScript is very low, so I will adjust it.

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -679,6 +679,17 @@ namespace ts {
         }
         else {
+            const { baseUrl, paths } = state.compilerOptions;
+            if (baseUrl && paths && startsWithSlash(moduleName) && getOwnKeys(paths).find(startsWithSlash)) {
```
I think we should move this condition into isExternalModuleName and just use the path matching result. The fallback to matching these as relative module name if this condition is satisfied seems sketchy. This would also make all the places that handle the relative/non relative module names work correctly including the resolution caches.

EECOLOR replied:
How do you propose we do that?

When I look at the code for `tryLoadModuleUsingBaseUrl` I see the exact same mechanism. We first try to resolve it using the alias, if that fails we fall back on other behavior. Here we are doing the same.

This is the code in `isExternalModuleNameRelative`:
```js
pathIsRelative(moduleName) || isRootedDiskPath(moduleName)
```
If we wanted to move the condition here we would need to adjust its signature. We can not just check for the `/`. We actually need to first match against a `paths` pattern and if that fails we need to load it as a 'relative' file.

If I don't explain it correctly please look at the test cases I added.

I you however have a concrete suggestion I am very open to hearing it.

EECOLOR replied:
I have moved the code into `tryLoadModuleUsingRootDirs` to mirror `tryLoadModuleUsingBaseUrl`. 

Thinking about it, we might want to make it a top-level check. Something like:

```js
if (baseUrl && paths && startsWithSlash(moduleName) && getOwnKeys(paths).find(startsWithSlash)) {
    if (state.traceEnabled) {
        trace(state.host, Diagnostics.baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1, baseUrl, moduleName);
        trace(state.host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
    }
    const resolved = tryLoadModuleUsingPaths(extensions, moduleName, baseUrl, paths, loader, /*onlyRecordFailures*/ false, state);
    if (resolved) {
        return resolved.value;
    }
}

if (!isExternalModuleNameRelative(moduleName)) {
    return tryLoadModuleUsingBaseUrl(extensions, moduleName, loader, state);
}
else {
    return tryLoadModuleUsingRootDirs(extensions, moduleName, containingDirectory, loader, state);
}
```

This however would be a backwards incompatible change with regards to the traces (not sure if that is a problem).

EECOLOR replied:
I just realized that moving the piece of code out of `tryLoadModuleUsingBaseUrl` and moving the added code out of the if would be potentially a performance hit because there are usually far more `./` imports which should not resolved using path mappings.

EECOLOR commented:
Question for the maintainers: do you guys want me to squash the commits once the code has stabilized?

EECOLOR commented:
@sheetalkamat I have made some changes:
- Added a test to reflect the intent expressed in the result of the design meeting
- Changed the names of the tests to better reflect the intent
- Adjusted the code to more clearly communicate the flow: "if we have paths available and it is eligible for path mapping, try to load using the paths"

EECOLOR commented:
There is one potential problem left.

When thinking about this problem it is important to remember that the function `isExternalModuleNameRelative` means `isExternalModuleNameRelativeOrRoot` or `isExternalModuleNameNotANodeModule'. And when you read `!isExternalModuleNameRelative(moduleName)` you could say `isExternalModuleNameANodeModule`.

Before this change, all `!isExternalModuleNameRelative(moduleName)` could potentially be mapped using paths. And, the other way around, mapped paths would always be non-root and non-relative.

With this change, all `!pathIsRelative(moduleName)` **and** all `isRootedDiskPath(moduleName) || isUrl(moduleName)` could potentially be mapped using paths (where the last case has the requirement that we have a potential mapping). And, the other way around, mapped paths would always be non-relative.

This is not necessarily a problem. However, in some code bases there is 'invisible' coupling between concepts. The coupling for this case would be the assumption that `!isExternalModuleNameRelative(moduleName)` is potentially mapped.

As far as I can tell (by tracing back the introduction of path mapping and the introduction of root path's as 'relative') any coupling that exist would be accidental. I can also imagine that if a developer used `!isExternalModuleNameRelative(moduleName)` to mean `a node module` he/she did not think about the possibility of it being a mapped path and thus potentially *not* a node module.

Looking at the uses of `isExternalModuleNameRelative` my domain knowledge is lacking to determine the following two things:
- Did it contain a wrong assumption about the meaning of `isExternalModuleNameRelative` (in combination with paths) before this change?
- Does this change impact correct assumptions about the meaning of `isExternalModuleNameRelative` (in combination with paths)?

My advise would be to create a separate issue to track down the uses of `isExternalModuleNameRelative` and:
- investigate if it has any relation to node modules or rooted paths
- determine if the fact that those paths might actually not be a node module or rooted path (once resolved) has any impact

EECOLOR commented:
I have squashed the commits.

EECOLOR commented:
@sheetalkamat What's next?

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -675,6 +675,13 @@ namespace ts {
         state: ModuleResolutionState): Resolved | undefined {
 
+        const { baseUrl, paths } = state.compilerOptions;
+        if (baseUrl && paths && isEligibleForPathMapping(moduleName, paths)) {
```
This should be pulled out into

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -761,22 +778,13 @@ namespace ts {
         if (state.traceEnabled) {
             trace(state.host, Diagnostics.baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1, baseUrl, moduleName);
         }
-        if (paths) {
```
I think this should be pulled out into separate function instead of inlining. Its easier to read.
We also need trace above this line about using baseUrl in that.

EECOLOR replied:
> I think this should be pulled out into separate function instead of inlining.

I thought about that. I however couldn't think of a name. The name `tryLoadModuleUsingPaths` is already taken. What do you suggest?

> We also need trace above this line about using baseUrl in that.

The trace `baseUrl_option_is_set_to_0_using_this_value_to_resolve_non_relative_module_name_1` is not exactly correct for what is happening. And since we have the trace `paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0` I thought that would be enough. But if you really want it there I could of course add it back in.

sheetalkamat replied:
Yes please add that back in there since its used in there anyways.

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -683,6 +690,16 @@ namespace ts {
+    function isEligibleForPathMapping(moduleName: string, paths: MapLike<string[]>): boolean {
+        if (isRootedDiskPath(moduleName) || isUrl(moduleName)) {
+            // check for potential alias otherwise { '*': './src/*' } will match
+            const root = moduleName.slice(0, getRootLength(moduleName));
```
You don't need all these checks. The path mapping needs to be done if: moduleName is not relative path or it is absolute path.

EECOLOR replied:
Did you read the comment?

It would break backwards compatibility and existing configurations.

sheetalkamat replied:
@DanielRosenwasser isn't this change expected breaking change ?

DanielRosenwasser replied:
Yes, I thought that was the point of this PR in the first place. What exactly are the extra checks meant to protect against?

EECOLOR replied:
Root (or url) imports failing (resolving to another file) in existing projects failing because they have a `*` mapping.

sheetalkamat replied:
As mentioned this is desired to be a breaking change. Please remove these extra checks

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -675,6 +675,13 @@ namespace ts {
+            if (resolved) return resolved.value;
+        }
+
         if (!isExternalModuleNameRelative(moduleName)) {
```
Also needs some tests to ensure that caching is working as intended (and hasn't changed)

EECOLOR replied:
Could you point me in the right direction? I am unsure what would be needed.

sheetalkamat replied:
There are tests in `src/testRunner/moduleResolution.ts` that verify caching is working as expected.

EECOLOR replied:
@sheetalkamat I figured it out, they are run by default

sheetalkamat commented:
Also need changes in `getStringLiteralCompletionsFromModuleNamesWorker` and corresponding tests with module name completions..

EECOLOR commented:
> Also need changes in `getStringLiteralCompletionsFromModuleNamesWorker` and corresponding tests with module name completions.

What kind of inputs would this fail on?

sheetalkamat commented:
More than failing of completions, we need to make sure the path completions are correct when there is path mapping.

EECOLOR commented:
@sheetalkamat I have updated the code to reflect your comments (apart from the ones that still have questions). I also squashed my commits again and rebased on master. Could you please take another look?

DanielRosenwasser commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -738,6 +741,24 @@ namespace ts {
+        if (isRootedDiskPath(moduleName) || isUrl(moduleName)) {
+            // check for potential alias otherwise { '*': './src/*' } will match
+            const root = moduleName.slice(0, getRootLength(moduleName));
+            const potentialAlias = getOwnKeys(paths).find(path => path.startsWith(root));
```
`find` and `startsWith` are ES2015 and we support ES5 environments, so use the helper functions from `core.ts`

DanielRosenwasser commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -738,6 +741,24 @@ namespace ts {
+            // check for potential alias otherwise { '*': './src/*' } will match
+            const root = moduleName.slice(0, getRootLength(moduleName));
+            const potentialAlias = getOwnKeys(paths).find(path => path.startsWith(root));
+            return Boolean(potentialAlias);
```
`!!`

sheetalkamat commented on src/services/stringCompletions.ts:
```diff
@@ -247,16 +247,17 @@ namespace ts.Completions.StringCompletions {
-            if (compilerOptions.rootDirs) {
-                return getCompletionEntriesForDirectoryFragmentWithRootDirs(
-                    compilerOptions.rootDirs, literalValue, scriptDirectory, extensionOptions, compilerOptions, host, scriptPath);
+        if (isRootedDiskPath(literalValue) || isUrl(literalValue)) {
```
@andy-ms can you please review completions change

andy-ms commented on src/services/stringCompletions.ts:
```diff
@@ -247,16 +247,17 @@ namespace ts.Completions.StringCompletions {
-                    compilerOptions.rootDirs, literalValue, scriptDirectory, extensionOptions, compilerOptions, host, scriptPath);
+        if (isRootedDiskPath(literalValue) || isUrl(literalValue)) {
+            if (compilerOptions.baseUrl) {
+                return getCompletionEntriesForNonRelativeModules(literalValue, scriptDirectory, compilerOptions, host, typeChecker);
```
The two `getCompletionEntriesForNonRelativeModules` calls are identical, as are the two `getCompletionEntriesForRelativeModules` calls. This could be a single conditional instead of several nested ones.
```ts
return isPathRelativeToScript(literalValue) || !compilerOptions.baseUrl && (isRootedDiskPath(literalValue) || isUrl(literalValue))
    ? getCompletionEntriesForRelativeModules(literalValue, scriptDirectory, compilerOptions, host, scriptPath)
    : getCompletionEntriesForNonRelativeModules(literalValue, scriptDirectory, compilerOptions, host, typeChecker);
```

EECOLOR commented:
@sheetalkamat I have made the required changes.

andy-ms commented:
@sheetalkamat Those are really difficult to revert -- it happens when the submodule is updated by git. I think it's fine because we regularly update these anyway.

sheetalkamat commented:
@EECOLOR thank you for contribution. This should be available in tonight's nightly build.

EECOLOR commented:
@sheetalkamat thank you and your team for your help

EECOLOR commented:
@sheetalkamat did something go wrong with the nightly build?

sheetalkamat commented:
@weswigham @DanielRosenwasser would know.

weswigham commented:
Looks like for some reason the nightly publish job didn't get scheduled until I just went to go check on the task. Likely an azure devops issue - since I believe it just triggered, we'll probably publish a nightly in about 30 minutes.
